polygon/p2p/downloader.go

```diff
@@ -0,0 +1,151 @@
+	if start > end {
+		return nil, fmt.Errorf("%w: start=%d, end=%d", invalidDownloadHeadersRangeErr, start, end)
+	}
+
+	var headers []*types.Header
+	reqId := d.requestIdGenerator()
+
+	observer := make(chanMessageObserver)
+	d.messageListener.RegisterBlockHeaders66(observer)
+	defer d.messageListener.UnregisterBlockHeaders66(observer)
+
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+	g, ctx := errgroup.WithContext(ctx)
+
+	g.Go(func() error {
+		for {
+			select {
+			case <-ctx.Done():
+				return fmt.Errorf("interrupted while waiting for msg from peer: %w", ctx.Err())
+			case msg := <-observer:
+				msgPid := PeerIdFromH512(msg.PeerId)
+				if msgPid != pid {
+					continue
+				}
+
+				var pkt eth.BlockHeadersPacket66
+				if err := rlp.DecodeBytes(msg.Data, &pkt); err != nil {
+					if rlp.IsInvalidRLPError(err) {
+						d.logger.Debug("penalizing peer for invalid rlp response", "pid", pid.String())
+						d.peerManager.Penalize(pid)
+					}
+
+					return fmt.Errorf("failed to decode BlockHeadersPacket66: %w", err)
+				}
+
+				if pkt.RequestId != reqId {
+					continue
+				}
+
+				//
+				// TODO what sort of validation if any should we do here? check sentry_multi_client & refactor if necessary
+				// Looks like sentry_multi_client does not actually do any validations
+				// corresponding to the spec at this layer (p2p message processing).
+				//
+				// We should make a decision which of the following validations we want to do here
+				// and which at higher layers.
+				//
+				// From the spec:
+				/*
+					The validity of block headers depends on the context in which they are used. For a single block header,
+					only the validity of the proof-of-work seal (mix-digest, block-nonce) can be verified. When a header is
+					used to extend the client's local chain, or multiple headers are processed in sequence during chain
+					synchronization, the following rules apply:
+
+					- Headers must form a chain where block numbers are consecutive and the parent-hash of each header
+					matches the hash of the preceding header.
+					- When extending the locally-stored chain, implementations must also verify that the values of
+					difficulty, gas-limit and time are within the bounds of protocol rules given in the Yellow Paper.
+					- The gas-used header field must be less than or equal to the gas-limit.
+					- basefee-per-gas must be present in headers after the London hard fork. It must be absent for earlier
+					blocks. This rule was added by EIP-1559.
+					- For PoS blocks after The Merge, ommers-hash must be the empty keccak256 hash since no ommer headers
+					can exist.
+					- withdrawals-root must be present in headers after the Shanghai fork. The field must be absent for
+					blocks before the fork. This rule was added by EIP-4895.
+				*/
+
+				//
+				// TODO also limit on size of message 10MiB and penalize?
+				// TODO should we check for too large of a range for download headers?
```

Comment:
@battlmonstr we should maybe discuss these validation related questions when you have some time. What are your thoughts? I'm planning on addressing these TODOs in a separate PR.

Reply:
yes, we should, in a separate PR 👌

Reply:
updated todos based on our conversation